c2rust-analyze/src/util.rs

```diff
@@ -353,3 +354,64 @@ pub fn is_null_const(constant: Constant) -> bool {
 
 pub trait PhantomLifetime<'a> {}
 impl<'a, T: ?Sized> PhantomLifetime<'a> for T {}
+
+/// Determine if `from` can be safely transmuted to `to`,
+/// which is defined as `*(from as *const To)` being a safe operation,
+/// where `from: *const From` and assuming `*from` already was safe.
+///
+/// Note that this is one-way, and is slightly different from [`core::mem::transmute`],
+/// and more similar to [`core::mem::transmute_copy`].
+///
+/// This forms a reflexive, transitive, and non-symmetric (one-way) relation, named `~` below.
+/// Formally, `A ~ B` iff `*a` and `*(a as *const B)` are safe, where `a: *const A`.
+///
+/// However, safe transmutability is difficult to check completely,
+/// so this function only checks a subset of it,
+/// with these formal rules for all types `A`, `B`:
+///
+/// * `A = B => A ~ B`
+/// * `A ~ B => *A ~ *B`
+/// * `uN ~ iN`, `iN ~ uN`, where `N` is an integer width
+/// * `[A] ~ A`
+/// * `[A; N] ~ A`, where `const N: usize`
+///
+/// Thus, [`true`] means it is definitely transmutable,
+/// while [`false`] means it may not be transmutable.
+pub fn is_transmutable_to<'tcx>(from: Ty<'tcx>, to: Ty<'tcx>) -> bool {
+    let transmutable_ints = || {
```

Comment:
I assume `transmutable_ints` and `one_way_transmutable` are closures so the check at the end can short-circuit to avoid them?  I suspect that's not necessary - LLVM should be smart enough to delay the memory loads as late as possible.

Reply:
> I assume `transmutable_ints` and `one_way_transmutable` are closures so the check at the end can short-circuit to avoid them?

Yes.

> I suspect that's not necessary - LLVM should be smart enough to delay the memory loads as late as possible.

`one_way_transmutable` is recursive, so I don't think this is simple to elide.  Besides, I'd rather not depend on LLVM being super clever as the closure is very simple (just adding `||` to definitions and `()`s to calls).